### ISSUE TITLE
Incorrect feature choice from unnormalized data using LIME Tabular

### ISSUE THREAD

**Issue by eddieh88** [marcotcr/lime#121]
Hi,
I am trying to understand in detail LIME explanations for tabular data. As I understand it, the choice of top features is done incorrectly. 

In line:
[Lime Tabular 1.](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L214-L215)

a generic mean of 0 and standard deviation of 1 is given for all features.  This leads to the line below doing nothing:

[Lime Tabular --  __data_inverse](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L261)

Reminder this 'data, inverse' coming from '__data_inverse' maps the normally distributed data to a binary matrix. Each row of this binary matrix (outside the first row) has a '1' for each column if the associated value is in the same quantile as the test row (first row). It is certainly not 'normalized'.

'scaled_data' is then passed to the LimeBase method 'explain_instance_with_data':

[Lime tabular -- explain_instance_with_data'](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L362)

And then to:

[Lime base -- 'explain_instance_with_data'](https://github.com/marcotcr/lime/blob/master/lime/lime_base.py#L108)

Finally this binary matrix is passed to:

[Lime base 'feature_selection'](https://github.com/marcotcr/lime/blob/master/lime/lime_base.py#L153-L153)

then:

[Lime base -- feature_selection](https://github.com/marcotcr/lime/blob/master/lime/lime_base.py#L70)

Each one of the sklearn Linear algorithms used in the function 'feature_selection' have as default normalize = False, so no normalizing is being done. However, we do use the size of the coefficients as a ranking for the attributes even though no normalization has been done:

[Lime base -- feature_selection 2. ](https://github.com/marcotcr/lime/blob/master/lime/lime_base.py#L81-L84)

This is a no-no in standard linear regression, [seen here](https://stats.stackexchange.com/questions/202221/for-linear-classifiers-do-larger-coefficients-imply-more-important-features/202846) I am sure there is a better (more academic) reference than this for feature importance in linear regression, [or here](http://blog.minitab.com/blog/adventures-in-statistics-2/how-to-identify-the-most-important-predictor-variables-in-regression-models)

My quick solution was to add:

```
self.scaler2 = sklearn.preprocessing.StandardScaler(with_mean=False)
        self.scaler2.fit(data)
        scaled_data = (data - self.scaler2.mean_) / self.scaler2.scale_
```

at [lines.](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L260-L261)

This is redundant to [here](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L214-L215)

One also to consider how the normalizing skews the distance between points just [below](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L263-L267)

But I assume these distances will be ok after normalization (looks to be the authors intent from the gecko). 

Let me know if my concerns are warranted,
Eddie

**Comment by echo66** [comment]
@eddieh88 

Not all features will have mean 0 and scale 1, only the categorical ones. Notice the parent for cycle.

**Comment by eddieh88** [comment]
@echo66 I filed this issue on the [iris example](https://marcotcr.github.io/lime/tutorials/Tutorial%20-%20continuous%20and%20categorical%20features.html)

Apologies, should have said that earlier. 

These features are initially numeric (just the standard 4 numeric features and the species), at some point when they are converted to a binary matrix they are converted to categorical? 

Thanks for the response,
Eddie

**Comment by echo66** [comment]
They are converted to categorical because that example is using a discretizer. Look at cell 6.

**Comment by eddieh88** [comment]
@echo66  Ok, I get that. But at the end of the day you have a discretized set you pass unnormalized with a continuous target variable to a linear regression algorithm. You then sort the coefficients from the regression by magnitude to get feature importance, correct? Is that ok with you? Thanks

**Comment by marcotcr** [comment]
Hey, sorry for the delayed response.
As @echo66 pointed out (thanks, btw), categorical features do not get scaled ([this](https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L214) is just acting on features that were discretized). If you have numerical features, they do get scaled.

In the iris example, after discretization what we end up passing to the linear regression model is a binary matrix, where you have a 1 if the feature value in the perturbed sample is within the same bucket as the feature value in example being explained, and 0 otherwise.
Does this make sense? If so, do you still object to looking at coefficients from the regression on this matrix?

Thanks,

**Comment by eddieh88** [comment]
@marcotcr 
```
In the iris example, after discretization what we end up passing to the linear regression model is a binary matrix, where you have a 1 if the feature value in the perturbed sample is within the same bucket as the feature value in example being explained, and 0 otherwise.
Does this make sense? 
```
Yes this makes sense. I guess I considered if one had a mix of of these binary features and numeric features, I would be concerned about using the size of the coefficients for rank importance. I am not 100% sure it is always ok with the purely discrete case. 

I have been reading Lundberg's Shapley paper (https://arxiv.org/abs/1705.07874) and he relates your LIME paper local linear expression with binary inputs to Shapley coefficients. From there, the coefficients are used to rank importance; normalization is not done throughout I am pretty sure. These Shapley coeffs are just treated as some kind of invariants to the local prediction of some point.  

So I guess I will just think on this issue for awhile and not cause you all any more trouble.

**Comment by marcotcr** [comment]
No trouble at all : )

**Comment by marcotcr** [comment]
Reopen the issue if you have further thoughts
